src/show_message.rs

```diff
@@ -0,0 +1,137 @@
+use std::borrow::Cow;
+
+use itertools::Itertools;
+use jsonrpc_core::{Id, MethodCall};
+use lsp_types::{MessageActionItem, MessageType, ShowMessageParams, ShowMessageRequestParams};
+use serde::Deserialize;
+
+use crate::{context::Context, types::EditorMeta, util::editor_quote};
+
+/// Queues the message request from the LSP server.
+pub fn show_message_request(meta: EditorMeta, request: MethodCall, ctx: &mut Context) {
+    let params: ShowMessageRequestParams = request
+        .params
+        .clone()
+        .parse()
+        .expect("Failed to parse ShowMessageRequest params");
+    ctx.pending_message_requests.push_back((request.id, params));
+    update_modeline(meta, ctx)
+}
+
+#[derive(Deserialize)]
+struct MessageRequestResponse {
+    pub message_request_id: Option<Id>,
+    pub item: Option<toml::Value>,
+}
+
+/// Handles an user's response to a message request (or the user's request to display the next message request).
+pub fn show_message_request_respond(meta: EditorMeta, params: toml::Value, ctx: &mut Context) {
+    let resp =
+        MessageRequestResponse::deserialize(params).expect("Cannot parse message request response");
+    let req_id = match resp.message_request_id {
+        None => return show_message_request_next(meta, ctx),
+        Some(id) => id,
+    };
+    let item = resp
+        .item
+        .and_then(|v| MessageActionItem::deserialize(v).ok())
+        .map(|v| jsonrpc_core::to_value(v).expect("Cannot serialize item"))
+        .unwrap_or(jsonrpc_core::Value::Null);
+    ctx.reply(req_id, Ok(item));
+}
+
+fn show_message_request_next(meta: EditorMeta, ctx: &mut Context) {
+    let (id, params) = match ctx.pending_message_requests.pop_front() {
+        Some(v) => v,
+        None => {
+            return ctx.exec(meta, "lsp-show-error 'No pending message requests.'");
+        }
+    };
+
+    let options = match &params.actions {
+        Some(opts) if !opts.is_empty() => &opts[..],
+        _ => {
+            // a ShowMessageRequest with no actions is just a ShowMessage notification.
+            show_message(
+                meta,
+                ShowMessageParams {
+                    typ: params.typ,
+                    message: params.message,
+                },
+                ctx,
+            );
+            ctx.reply(id, Ok(serde_json::Value::Null));
+            return;
+        }
+    };
+    let request_id = editor_quote(
+        toml::to_string(&id)
+            .expect("cannot convert request id to toml")
+            .as_ref(),
+    );
+    let option_menu_opts = options
+        .iter()
+        .flat_map(|item| {
+            let cmd = format!(
+                "%{{lsp-show-message-request-respond {} {}}}",
```

Comment:
don't use nested quoting for user input because we don't know if `item` has balanced quotes.
It should be like

	editor_quote(&format!("lsp-show-message-request-respond {} {}", ...))